graph/src/data/graphql/validation.rs

```diff
@@ -93,10 +100,199 @@ pub fn get_object_type_definitions(schema: &Document) -> Vec<&ObjectType> {
         .collect()
 }
 
+/// Returns all object and interface type definitions in the schema.
+pub fn get_object_and_interface_type_fields(schema: &Document) -> HashMap<&Name, &Vec<Field>> {
+    schema
+        .definitions
+        .iter()
+        .filter_map(|d| match d {
+            Definition::TypeDefinition(TypeDefinition::Object(t)) => Some((&t.name, &t.fields)),
+            Definition::TypeDefinition(TypeDefinition::Interface(t)) => Some((&t.name, &t.fields)),
+            _ => None,
+        })
+        .collect()
+}
+
 /// Looks up a directive in a object type, if it is provided.
 pub fn get_object_type_directive(object_type: &ObjectType, name: Name) -> Option<&Directive> {
     object_type
         .directives
         .iter()
         .find(|directive| directive.name == name)
 }
+
+/// Returns the underlying type for a GraphQL field type
+pub fn get_base_type(field_type: &Type) -> &Name {
+    match field_type {
+        Type::NamedType(name) => name,
+        Type::NonNullType(inner) => get_base_type(&inner),
+        Type::ListType(inner) => get_base_type(&inner),
+    }
+}
+
+/// Check `@derivedFrom` annotations for various problems. This follows the
+/// corresponding checks in graph-cli
+fn validate_derived_from(schema: &Document) -> Result<(), SchemaValidationError> {
+    // Helper to construct a DerivedFromInvalid
+    fn invalid(object_type: &ObjectType, field_name: &str, reason: &str) -> SchemaValidationError {
+        SchemaValidationError::DerivedFromInvalid(
+            object_type.name.to_owned(),
+            field_name.to_owned(),
+            reason.to_owned(),
+        )
+    }
+
+    let type_definitions = get_object_type_definitions(schema);
+    let object_and_interface_type_fields = get_object_and_interface_type_fields(schema);
+
+    // Iterate over all derived fields in all entity types; include the
+    // `field` argument of @derivedFrom directive
+    for (object_type, field, target_field) in type_definitions
```

Comment:
We should probably move the schema validation code into the `graph-graphql` crate. Like you already mentioned, a number of helpers have already been added there but can't be used. This is another: `graph_graphql::schema::ast::get_derived_from_directive` and `graph_graphql::schema::ast::get_derived_from_field` already exist there.

Reply:
We don't have to do it now though.

Reply:
The alternative would be to make the `graph_graphql` crate unaware of an Graph-specific data structures and make it a dependency of `graph`.

Reply:
I agree that that would be good to do, but I also do not want to do that as part of this PR which just builds on what's there already. I opened #1265 to track that.